### DISCUSSION TITLE
Porting Doom to ELKS :)

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1343] [Ideas]
Hi,

It seems DOOM can be used with direct VGA access and this theoretically opens the door for porting it to ELKS when using a 386. It already works on Xenix and A/UX.

https://www.osnews.com/story/134966/porting-doom-to-a-ux/
https://katelibc.medium.com/porting-doom-to-a-ux-8cecab02b531

Full source code of the port to Xenix:
https://github.com/gattilorenz/Xenix-doom

**Comment by ghaerr** [comment]
I ported DOOM to nano-X some time ago, source code is here: https://github.com/ghaerr/microwindows/tree/master/src/contrib/doom. 
Nano-X is ported to ELKS and is using the EGA/VGA driver.

On ELKS the nano-X server and application are combined into a single executable, so medium model would need to be used, which hopefully is enough for both with a max code size of 128k. On older machines, DOOM could be too slow to be playable, due to its intensive graphics and the very slow VGA driver. I've only tested it on 32-bit Linux with framebuffer.

**Reply by toncho11** [reply]
Your code will generate a single executable that contains both nano-X and Doom?
We need to try to compile with gcc-16 used by ELKS.
We need to set medium memory model in gcc-16 used by ELKS.

And then it can be tested on any 386 with VGA card.

**Comment by ghaerr** [comment]
> Your code will generate a single executable that contains both nano-X and Doom?

Yes, the version in the Microwindows Project can be configured to do that.

> We need to try to compile with gcc-16 used by ELKS.
We need to set medium memory model in gcc-16 used by ELKS.

Yep, contributions welcome :) I've got my hands full but glad to help with technical questions.

**Comment by toncho11** [comment]
Here is an old version of DOOM that could be ported to ELKS using [SVGALib](https://www.svgalib.org/):

https://sourceforge.net/projects/lxdoom/files/lxdoom/1.4.4/

**Comment by ghaerr** [comment]
See https://github.com/jbruchon/elks/discussions/1447#discussioncomment-4161440 for more comments on 16-bit versus 32-bit ELKS compiler, SVGALib and Doom support.

**Comment by FrenkelS** [comment]
https://github.com/FrenkelS/elksdoom
The main menu works. Loading a level gives an out of memory error on my machine. :(

**Comment by toncho11** [comment]
Currently elksdoom does not crash because of insufficient memory! It works quite well!
Check this release: https://github.com/FrenkelS/elksdoom/issues/2#issuecomment-2331922934
Or download the latest. In this case you need to enable certain memory optimizations that make ELKS use less memory which is in turn freed for doomelks.
